partition.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright (c) 2024 Diagrid Inc.
+Licensed under the MIT License.
+*/
+
+package etcdcron
+
+import (
+	"fmt"
+	"hash/fnv"
+)
+
+// The Partitioning splits the jobs between partitions.
+type Partitioning interface {
+	CalculatePartitionId(key string) int
+	ListPartitions() []int
+	CheckPartitionLeader(partitionId int) bool
+}
+
+type partitioning struct {
+	// Number of hosts to compete to leadership of partitions.
+	numHosts int
+	// Maximum number of virtual partitions to split the jobs.
+	numVirtualPartitions int
+	// This hostId (must be between 0 and numHosts)
+	hostId int
+	// Partitions that this instance will try to become a leader for.
+	myPartitions []int
+}
+
+// NoPartitioning will make any host a leader of the only partition.
+func NoPartitioning() Partitioning {
+	p, _ := NewPartitioning(1, 1, 0)
+	return p
+}
+
+// NewPartitioning holds the logic to hold job partitioning
+func NewPartitioning(numVirtualPartitions int, numHosts int, hostId int) (Partitioning, error) {
+	if numVirtualPartitions < numHosts {
+		return nil, fmt.Errorf("cannot have more hosts than partitions")
+	}
+	if hostId < 0 {
+		return nil, fmt.Errorf("hostId cannot be zero")
+	}
+	if hostId >= numHosts {
+		return nil, fmt.Errorf("hostId cannot be greater or equal to %d", numHosts)
+	}
+	partitions := []int{}
+	for i := 0; i < numVirtualPartitions; i++ {
+		if checkPartitionLeader(i, numVirtualPartitions, hostId, numHosts) {
+			partitions = append(partitions, i)
+		}
+	}
+	return &partitioning{
+		numVirtualPartitions: numVirtualPartitions,
+		numHosts:             numHosts,
+		hostId:               hostId,
+		myPartitions:         partitions,
+	}, nil
+}
+
+func (p *partitioning) CalculatePartitionId(key string) int {
+	// Changing this algorithm is a breaking change and stored data is no longer valid.
+	hash := fnv.New32a()
+	hash.Write([]byte(key))
+	return int(uint32(hash.Sum32()) % uint32(p.numVirtualPartitions))
+}
+
+func (p *partitioning) ListPartitions() []int {
+	return p.myPartitions
+}
+
+func (p *partitioning) CheckPartitionLeader(partitionId int) bool {
+	return checkPartitionLeader(partitionId, p.numVirtualPartitions, p.hostId, p.numHosts)
+}
+
+func checkPartitionLeader(partitionId, numVirtualPartitions, hostId, numHosts int) bool {
+	// First, make sure partitionId is within range.
+	sanitizedPartitionId := positiveMod(partitionId, numVirtualPartitions)
+	// Now, check if the mods match.
+	return positiveMod(hostId, numHosts) == positiveMod(sanitizedPartitionId, numHosts)
+}
+
+func positiveMod(key, max int) int {
+	result := key % max
+	if result < 0 {
```

Comment:
key can be negative, in theory.

Reply:
Could it be negative _more_ than max?

Reply:
It can be any negative number valid within the int range.

Reply:
This is a private method, not supposed to be used directly. NewPartitioning() has validation to avoid negative numbers. At this point, positiveMod is not needed anymore. Let me remove it.

```
	if numVirtualPartitions < numHosts {
		return nil, fmt.Errorf("cannot have more hosts than partitions")
	}
	if hostId < 0 {
		return nil, fmt.Errorf("hostId cannot be zero")
	}
	if hostId >= numHosts {
		return nil, fmt.Errorf("hostId cannot be greater or equal to %d", numHosts)
	}
```

Reply:
Keeping it because there is one input int that is not validated and can be negative:

```
func (p *partitioning) CheckPartitionLeader(partitionId int) bool {
	return checkPartitionLeader(partitionId, p.numVirtualPartitions, p.hostId, p.numHosts)
}
```

But max cannot be negative.